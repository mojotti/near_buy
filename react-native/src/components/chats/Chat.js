import React from 'react';
import { View } from 'react-native';


export default class Chat extends React.Component {
  static navigationOptions = ({ navigation }) => {
    return { title: navigation.state.params.item.title };
  };

  render() {
    console.log('thos', this.props);
    return (
      <View />
    );
  }
}
