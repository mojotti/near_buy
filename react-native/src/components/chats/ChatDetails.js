import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Dimensions, Text, TouchableOpacity, View } from 'react-native';
import { connect } from 'react-redux';
import { withNavigation } from 'react-navigation';
import { styles } from '../../static/styles/ListViewItemStyle';
import { localhost } from '../../static/constants';
import ImagePlaceholder from '../image_placeholder/ImagePlaceholder';

const WIDTH = Dimensions.get('window').width;

export class _ChatDetails extends PureComponent {
  _navigateToChat = () => {
    console.log('navigation', this.props)
    const { navigate } = this.props.navigation;
    navigate('Chat', { item: this.props.item.item });
  };

  render() {
    const id = this.props.item.item.item_id;
    const imagePath = `http://${localhost}:5000/api/v1.0/${id}/image0.jpg`;

    return (
      <View>
        <TouchableOpacity onPress={this._navigateToChat}>
          <View style={styles.itemContainer}>
            <View style={styles.imagePlaceholderContainer}>
              <View style={styles.imageContainer}>
                <ImagePlaceholder
                  styles={styles.image}
                  url={imagePath}
                  placeholderSize={WIDTH * 0.15}
                />
              </View>
            </View>
            <View style={styles.item}>
              <Text style={styles.title}>{this.props.item.item.title}</Text>
            </View>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
}

_ChatDetails.propTypes = {
  item: PropTypes.shape({
    item: PropTypes.shape({
      item_id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      seller_id: PropTypes.number.isRequired,
      buyer_id: PropTypes.number.isRequired,
    }).isRequired,
  }).isRequired,
};

const mapStateToProps = (state, ownProps) => {
  return {};
};

const ChatDetails = connect(mapStateToProps, null)(_ChatDetails);
export default withNavigation(ChatDetails);
