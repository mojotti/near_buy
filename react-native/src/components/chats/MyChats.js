import React from 'react';
import { FlatList, TouchableOpacity, View } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/MaterialIcons';

import MaterialIconAndText from '../common/MaterialIconAndText';
import {
  NO_CHATS_TEXT_HEADER,
  NO_CHATS_TEXT_CONTENT,
} from '../../static/constants';
import { requestChats } from '../../redux/actions/ChatActions';
import Chat from './ChatDetails';
import ItemSeparator from '../user_items/ItemSeparator';

export class _MyChats extends React.Component {
  static navigationOptions = ({ navigation }) => {
    return {
      title: 'Chats',
      headerLeft: (
        <View style={{ paddingHorizontal: 10 }}>
          <TouchableOpacity onPress={() => navigation.navigate('DrawerOpen')}>
            <Icon name="menu" size={30} color="#4d4dff" />
          </TouchableOpacity>
        </View>
      ),
    };
  };

  componentWillMount() {
    this._updateChats();
  }

  _updateChats = () => this.props.dispatch(requestChats(this.props.token));

  _renderNoChats = () => {
    return (
      <View style={{ flex: 1 }}>
        <MaterialIconAndText
          iconName="emoticon-poop"
          headerText={NO_CHATS_TEXT_HEADER}
          containerText={NO_CHATS_TEXT_CONTENT}
        />
      </View>
    );
  };

  _renderChat = rowData => <Chat item={rowData} itemId={rowData.index} />;

  _renderSeparator = () => <ItemSeparator widthPercentage={0.93} />;

  _renderChats = () => {
    return (
      <FlatList
        data={this.props.chatHeaders}
        renderItem={this._renderChat}
        keyExtractor={(item, index) => index.toString()}
        onRefresh={this._updateChats}
        refreshing={this.props.isFetching}
        ItemSeparatorComponent={this._renderSeparator}
      />
    );
  };

  render() {
    const isVisible =
      this.props.chatHeaders && this.props.chatHeaders.length > 0;
    const backgroundColor = isVisible ? '#FFFFFF' : 'transparent';
    const bgColor = { backgroundColor };

    return (
      <View style={[bgColor, { flex: 1 }]}>
        {isVisible ? this._renderChats() : this._renderNoChats()}
      </View>
    );
  }
}

_MyChats.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  token: PropTypes.string.isRequired,
  chatHeaders: PropTypes.arrayOf(
    PropTypes.shape({
      seller_id: PropTypes.number.isRequired,
      buyer_id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      item_id: PropTypes.number.isRequired,
    }).isRequired
  ).isRequired,
};

const mapStateToProps = state => {
  const { chatHeaders, isFetching } = state.currentChatsReducer;
  const { token } = state.authorizationReducer;

  return { chatHeaders, isFetching, token };
};

const MyChats = connect(mapStateToProps)(_MyChats);
export default MyChats;
