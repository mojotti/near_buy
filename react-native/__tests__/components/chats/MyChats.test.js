import React from 'react';
import { shallow } from 'enzyme';
import { _MyChats } from '../../../src/components/chats/MyChats';

const HEADERS = [{
  seller_id: 0,
  buyer_id: 1,
  title: 'foo',
  id: 0,
}];

describe('<_MyChats />', () => {
  test('shows placeholder when no items', () => {
    const wrapper = shallow(
      <_MyChats
        isFetching={false}
        token="foo"
        chatHeaders={[]}
        dispatch={() => {}}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('shows items when there are items', () => {
    const wrapper = shallow(
      <_MyChats
        isFetching={false}
        token="foo"
        chatHeaders={HEADERS}
        dispatch={() => {}}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('update calls correct actions', async () => {
    fetch.mockResponseSuccess(HEADERS);
    const dispatch = jest.fn();

    const wrapper = shallow(
      <_MyChats
        isFetching={false}
        token="foo"
        chatHeaders={HEADERS}
        dispatch={dispatch}
      />
    );

    await wrapper.instance()._updateChats();
    expect(dispatch.mock.calls).toMatchSnapshot();
  });

  test('renders navigation options correctly', () => {
    const navigation = {
      navigate: () => {}
    };
    expect(_MyChats.navigationOptions({ navigation })).toMatchSnapshot();
  });
});
