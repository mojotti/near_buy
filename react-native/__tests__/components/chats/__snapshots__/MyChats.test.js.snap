// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<_MyChats /> renders navigation options correctly 1`] = `
Object {
  "headerLeft": <View
    style={
      Object {
        "paddingHorizontal": 10,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
    >
      <Icon
        allowFontScaling={false}
        color="#4d4dff"
        name="menu"
        size={30}
      />
    </TouchableOpacity>
  </View>,
  "title": "Chats",
}
`;

exports[`<_MyChats /> shows items when there are items 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <FlatList
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "buyer_id": 1,
          "id": 0,
          "seller_id": 0,
          "title": "foo",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    onRefresh={[Function]}
    refreshing={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`<_MyChats /> shows placeholder when no items 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <MaterialIconAndText
      containerText="Start to socialize and get some cool stuff =)"
      headerText="You have no chats (yet)."
      iconName="emoticon-poop"
    />
  </View>
</View>
`;

exports[`<_MyChats /> update calls correct actions 1`] = `
Array [
  Array [
    [Function],
  ],
  Array [
    [Function],
  ],
]
`;
